stan/math/prim/fun/dot_self.hpp

```diff
@@ -21,14 +29,14 @@ inline double dot_self(const std::vector<double>& x) {
 /**
  * Returns the dot product of the specified vector with itself.
  *
- * @tparam R number of rows, can be Eigen::Dynamic
- * @tparam C number of columns, can be Eigen::Dynamic
+ * @tparam EigVec A type deriving from `Eigen::MatrixBase` with compile time
+ *  rows or columns equal to 1.
  * @param v Vector.
  * @throw std::domain_error If v is not vector dimensioned.
  */
-template <int R, int C>
-inline double dot_self(const Eigen::Matrix<double, R, C>& v) {
-  check_vector("dot_self", "v", v);
+template <typename EigMat,
+          require_eigen_vt<std::is_arithmetic, EigMat>* = nullptr>
+inline double dot_self(EigMat&& v) {
```

Comment:
I'm not reviewing, but would very much appreciate seeing this renamed to  `EigVec` (what I'd really like is `V`, but that's a bigger discussion we need to take up).  

It is *not* an Eigen matrix, it's an Eigen *vector*.  `Eigen::Matrix<T, 1, -1>` and `Eigen::Matrix<T, -1, 1>` and `Eigen::Matrix<T, -1, -1>` are three completely independent specializations of `Eigen::Matrix<T, R, C>`.  Only the last of the tree is a matrix.  The other two are vectors and do not even support `T& operator()(ptrdiff_t, ptrdiff_t);`.

Reply:
I find it weird, but there are even some test checking that `dot_self` works with (non-vector) matrices.